### PR TITLE
Create v8Schema with series index

### PR DESCRIPTION
**Pull request by aaron7**
First step for #416.

This index will be used to fetch series from the index. It will replace the metric name index and we will make the switch after the next steps are complete.

Decided to encode the `model.Metric` as JSON because it already implements the `json.Unmarshaler` interface.

Used sha256 for identifying the series as the variant of fnv64a which if used for fingerprints is not unique enough for indexing and using sha256 keeps it simple (we don't have to deal with collision logic).

**Test plan**
- Tests pass

### CODE REVIEW

**Comment by tomwilkie** [comment]
> Used the metric fingerprint for identifying the series. It uses a variant of fnv64a https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function. Is this suitable for use in our index?

As discussed, we're going to go with a sha256.

File: pkg/chunk/inmemory_storage_client.go [hunk 1]
```diff
@@ -135,9 +135,10 @@ func (m *MockStorage) BatchWrite(_ context.Context, batch WriteBatch) error {
 			items = append(items, mockItem{})
 			copy(items[i+1:], items[i:])
 		} else {
-			// Return error if duplicate write and not metric name entry
+			// Return error if duplicate write and not metric name entry or series entry
 			itemComponents := decodeRangeKey(items[i].rangeValue)
-			if !bytes.Equal(itemComponents[3], metricNameRangeKeyV1) {
+			if !bytes.Equal(itemComponents[3], metricNameRangeKeyV1) &&
+				!bytes.Equal(itemComponents[3], seriesRangeKeyV1) {
 				return fmt.Errorf("Dupe write")
```
Comment (tomwilkie) [inline pkg/chunk/inmemory_storage_client.go]: We should change this to error if its a duplicate within the batch, to match server behaviour.

File: pkg/chunk/schema.go [hunk 2]
```diff
@@ -536,3 +547,33 @@ func (v7Entries) GetReadQueries(bucket Bucket) ([]IndexQuery, error) {
...
+		return nil, err
+	}
+
+	fingerprintBytes := make([]byte, 8)
+	binary.LittleEndian.PutUint64(fingerprintBytes, uint64(labels.Fingerprint()))
+
+	seriesBytes, err := json.Marshal(labels)
+	if err != nil {
+		return nil, err
+	}
```
Comment (tomwilkie) [inline pkg/chunk/schema.go]: Can we just use Metric.String()?

https://github.com/prometheus/common/blob/master/model/metric.go#L54

File: pkg/chunk/inmemory_storage_client.go (hunk 1 above)
Comment (aaron7) [inline pkg/chunk/inmemory_storage_client.go]: We are already doing so a few lines above. I didn't remove the Dupe write check across all writes as it's still useful for checking if chunks are no duplicated.

File: pkg/chunk/schema.go (hunk 2 above)
Comment (aaron7) [inline pkg/chunk/schema.go]: Can we decode the string back into `model.Metric`?

Comment (tomwilkie) [inline pkg/chunk/schema.go]: I couldn't find a convenient function, no.

I don't like the idea of putting JSON in here, its a PITA to deal with backwards compatibility etc.  But I guess it'll have to do for now.

**Comment by aaron7** [comment]
#442

**Pull request closed without merging**